File: CODE/TEXT/DART/lib/parsing.dart
// -- IMPORTS

import 'dart:convert';
import 'constant.dart';
import 'processing.dart';

// -- CONSTANTS

final RegExp
    _decimalRealExpression = RegExp( r'^-?\d+(\.\d+)?$' ),
    _exponentialDecimalRealExpression = RegExp( r'^-?\d+(\.\d+)?[eE][-+]?\d+$' ),
    _hexadecimalIntegerExpression = RegExp( r'^0x[0-9A-Fa-f]+$' );

// -- TYPES

class DefLine
{
    // -- ATTRIBUTES

    final String
        line;
    final int
        lineSpaceCount;

    // -- CONSTRUCTORS

    DefLine(
        {
            required this.line,
            required this.lineSpaceCount
        }
        );
}

// ~~

class ParsingContext
{
    // -- ATTRIBUTES

    final String
        baseFolderPath,
        filePath;
    final String Function( String, [String] )?
        fileReadingFunction;
    final String
        stringProcessingQuote;
    final dynamic Function( String, ParsingContext, int )?
        stringProcessingFunction;
    final int
        levelSpaceCount;
    final String
        text;
    final List<String>
        lineArray;
    int
        lineIndex;

    // -- CONSTRUCTORS

    ParsingContext(
        {
            required this.baseFolderPath,
            required this.filePath,
            required this.fileReadingFunction,
            required this.stringProcessingQuote,
            required this.stringProcessingFunction,
            required this.levelSpaceCount,
            required this.text,
            required this.lineArray,
            this.lineIndex = 0,
        }
        );
}

// -- FUNCTIONS

String getKeyText(
    dynamic key
    )
{
    if ( key == undefined )
    {
        return 'undefined';
    }
    else if ( key is Map
              || key is List
              || key is bool
              || key is num )
    {
        return jsonEncode( key );
    }
    else
    {
        return key.toString();
    }
}

// ~~

List<String> getTokenArray(
    String text
    )
{
    var tokenArray = <String>[];
    var characterIndex = 0;

    while ( characterIndex < text.length )
    {
        if ( text[ characterIndex ] == '\\' )
        {
            if ( characterIndex + 1 < text.length )
            {
                tokenArray.add( text.substring( characterIndex, characterIndex + 2 ) );
                characterIndex += 2;
            }
            else
            {
                tokenArray.add( text[ characterIndex ] );
                ++characterIndex;
            }
        }
        else
        {
            var postCharacterIndex = characterIndex;

            while ( postCharacterIndex < text.length
                    && text[ postCharacterIndex ] != '\\' )
            {
                ++postCharacterIndex;
            }

            tokenArray.add( text.substring( characterIndex, postCharacterIndex ) );
            characterIndex = postCharacterIndex;
        }
    }

    return tokenArray;
}

// ~~

String getUnescapedText(
    List<String> tokenArray
    )
{
    var unescapedText = '';

    for ( var token in tokenArray )
    {
        if ( token.length == 2
             && token[ 0 ] == '\\' )
        {
            switch ( token[ 1 ] )
            {
                case 'n':
                    {
                        unescapedText += '\n';

                        break;
                    }

                case 't':
                    {
                        unescapedText += '\t';

                        break;
                    }

                case 'r':
                    {
                        unescapedText += '\r';

                        break;
                    }

                case 'b':
                    {
                        unescapedText += '\b';

                        break;
                    }

                case 'f':
                    {
                        unescapedText += '\f';

                        break;
                    }

                case '0':
                    {
                        unescapedText += '\0';

                        break;
                    }

                default:
                    {
                        unescapedText += token[ 1 ];
                    }
            }
        }
        else
        {
            unescapedText += token;
        }
    }

    return unescapedText;
}

// ~~

void throwParsingError(
    String message,
    ParsingContext context,
    int level
    )
{
    message =
        message
        + '\nText :\n'
        + context.text
        + '\nFile : ' + context.filePath
        + '\nLine ' + context.lineIndex.toString() + ' @ ' + level.toString();

    if ( context.lineIndex > 0
         && context.lineIndex <= context.lineArray.length )
    {
        message += ' : ' + context.lineArray[ context.lineIndex - 1 ];
    }

    throw Exception( message );
}

// ~~

DefLine parseDefLine(
    ParsingContext context,
    int level
    )
{
    var line = context.lineArray[ context.lineIndex ];
    var trimmedLine = line.trimLeft();
    var levelSpaceCount = level * context.levelSpaceCount;
    var lineSpaceCount = line.length - trimmedLine.length;

    if ( trimmedLine == '' )
    {
        line = '';
        lineSpaceCount = 0;
    }
    else
    {
        if ( lineSpaceCount < levelSpaceCount )
        {
            throwParsingError( 'Invalid DEF line', context, level );
        }

        line = line.substring( levelSpaceCount ).trimRight();
        lineSpaceCount -= levelSpaceCount;
    }

    context.lineIndex++;

    return DefLine( line: line, lineSpaceCount: lineSpaceCount );
}

// ~~

String parseDefUnquotedString(
    ParsingContext context,
    int level
    )
{
    var string = '';

    while ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        var tokenArray = getTokenArray( line );
        var lastToken = tokenArray.isNotEmpty ? tokenArray.last : '';

        if ( lastToken == '\\' )
        {
            tokenArray.removeLast();
            string += getUnescapedText( tokenArray );
        }
        else
        {
            if ( line.endsWith( '¨' )
                 && lastToken != '\\¨' )
            {
                tokenArray[ tokenArray.length - 1 ] = lastToken.substring( 0, lastToken.length - 1 );
            }

            string += getUnescapedText( tokenArray );

            return string;
        }
    }

    throwParsingError( 'Invalid DEF unquoted string', context, level );

    return '';
 }

// ~~

dynamic parseDefQuotedString(
    ParsingContext context,
    int level
    )
{
    var firstLineIndex = context.lineIndex + 1;
    var string = '';
    var quote = '';
    var escapedQuote = '';

    while ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        if ( context.lineIndex == firstLineIndex )
        {
            quote = line[ 0 ];
            line = line.substring( 1 );
            escapedQuote = '\\' + quote;
        }

        var tokenArray = getTokenArray( line );
        var lastToken = tokenArray.isNotEmpty ? tokenArray.last : '';

        if ( lastToken == '\\' )
        {
            tokenArray.removeLast();

            string += getUnescapedText( tokenArray );
        }
        else if ( lastToken.endsWith( quote )
                  && lastToken != escapedQuote )
        {
            tokenArray[ tokenArray.length - 1 ] = lastToken.substring( 0, lastToken.length - 1 );

            string += getUnescapedText( tokenArray );

            if ( quote == context.stringProcessingQuote
                 && context.stringProcessingFunction != null )
            {
                return context.stringProcessingFunction!( string, context, level );
            }
            else
            {
                return string;
            }
        }
        else
        {
            if ( lastToken.endsWith( '¨' )
                 && lastToken != '\\¨' )
            {
                tokenArray[ tokenArray.length - 1 ] = lastToken.substring( 0, lastToken.length - 1 );
            }

            string += getUnescapedText( tokenArray ) + '\n';
        }
    }

    throwParsingError( 'Invalid DEF quoted string', context, level );

    return '';
 }

// ~~

List<dynamic> parseDefArray(
    ParsingContext context,
    int level
    )
{
    var array = <dynamic>[];

    while ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        if ( lineSpaceCount == 0
             && line == ']' )
        {
            return array;
        }
        else if ( lineSpaceCount == 0
                  && line == ']{}' )
        {
            if ( array.length == 0 )
            {
                return [];
            }
            else
            {
                if ( array[ 0 ] is List )
                {
                    if ( array.length == 1 )
                    {
                        return [];
                    }
                    else
                    {
                        var keyArray = <String>[];

                        for ( var key in array[ 0 ] )
                        {
                            keyArray.add( getKeyText( key ) );
                        }

                        var keyCount = keyArray.length;
                        var objectArray = <Map<String, dynamic>>[];
                        var objectCount = array.length - 1;

                        for ( var objectIndex = 0;
                              objectIndex < objectCount;
                              ++objectIndex )
                        {
                            var valueArray = array[ objectIndex + 1 ];

                            if ( valueArray is List
                                 && valueArray.length == keyCount )
                            {
                                var object = <String, dynamic>{};

                                for ( var keyIndex = 0;
                                      keyIndex < keyCount;
                                      ++keyIndex )
                                {
                                    object[ keyArray[ keyIndex ] ] = valueArray[ keyIndex ];
                                }

                                objectArray.add( object );
                            }
                            else
                            {
                                throwParsingError( 'Invalid DEF object array', context, level );
                            }
                        }

                        return objectArray;
                    }
                }
                else
                {
                    throwParsingError( 'Invalid DEF object array', context, level );
                }
            }
        }
        else if ( lineSpaceCount == 0
                  && line == ']()' )
        {
            if ( array.length == 0 )
            {
                return [];
            }
            else
            {
                if ( array[ 0 ] is List )
                {
                    if ( array.length == 1 )
                    {
                        return [];
                    }
                    else
                    {
                        var keyArray = <String>[];

                        for ( var key in array[ 0 ] )
                        {
                            keyArray.add( getKeyText( key ) );
                        }

                        var keyCount = keyArray.length;
                        var mapArray = <Map<dynamic, dynamic>>[];
                        var mapCount = array.length - 1;

                        for ( var mapIndex = 0;
                              mapIndex < mapCount;
                              ++mapIndex )
                        {
                            var valueArray = array[ mapIndex + 1 ];

                            if ( valueArray is List
                                 && valueArray.length == keyCount )
                            {
                                var map = <dynamic, dynamic>{};

                                for ( var keyIndex = 0;
                                      keyIndex < keyCount;
                                      ++keyIndex )
                                {
                                    map[ keyArray[ keyIndex ] ] = valueArray[ keyIndex ];
                                }

                                mapArray.add( map );
                            }
                            else
                            {
                                throwParsingError( 'Invalid DEF map array', context, level );
                            }
                        }

                        return mapArray;
                    }
                }
                else
                {
                    throwParsingError( 'Invalid DEF map array', context, level );
                }
            }
        }
        else
        {
            context.lineIndex--;

            var value = parseDefValue( context, level + 1 );
            array.add( value );
        }
    }

    throwParsingError( 'Invalid DEF array', context, level );

    return [];
 }

// ~~

Map<String, dynamic> parseDefObject(
    ParsingContext context,
    int level
    )
{
    var object = <String, dynamic>{};

    while ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        if ( lineSpaceCount == 0
             && line == '}' )
        {
            return object;
        }
        else
        {
            context.lineIndex--;

            var key = parseDefValue( context, level + 1 );
            var value = parseDefValue( context, level + 2 );

            object[ getKeyText( key ) ] = value;
        }
    }

    throwParsingError( 'Invalid DEF object', context, level );

    return {};
 }

// ~~

Map<dynamic, dynamic> parseDefMap(
    ParsingContext context,
    int level
    )
{
    var map = <dynamic, dynamic>{};

    while ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        if ( lineSpaceCount == 0
             && line == ')' )
        {
            return map;
        }
        else
        {
            context.lineIndex--;

            var key = parseDefValue( context, level + 1 );
            var value = parseDefValue( context, level + 2 );

            map[ key ] = value;
        }
    }

    throwParsingError( 'Invalid DEF map', context, level );

    return {};
}

// ~~

dynamic parseDefValue(
    ParsingContext context,
    int level
    )
{
    if ( context.lineIndex < context.lineArray.length )
    {
        var DefLine( :line, :lineSpaceCount )
            = parseDefLine( context, level );

        if ( line == '[' )
        {
            return parseDefArray( context, level );
        }
        else if ( line == '{' )
        {
            return parseDefObject( context, level );
        }
        else if ( line == '(' )
        {
            return parseDefMap( context, level );
        }
        else if ( line.startsWith( '\'' )
                  || line.startsWith( '"' )
                  || line.startsWith( '`' )
                  || line.startsWith( '´' ) )
        {
            context.lineIndex--;

            return parseDefQuotedString( context, level );
        }
        else
        {
            var tokenArray = getTokenArray( line );

            if ( tokenArray.length == 1
                 && tokenArray[ 0 ] == line.trim() )
            {
                if ( line == 'undefined' )
                {
                    return undefined;
                }
                else if ( line == 'null' )
                {
                    return null;
                }
                else if ( line == 'false' )
                {
                    return false;
                }
                else if ( line == 'true' )
                {
                    return true;
                }
                else if ( line == 'NaN' )
                {
                    return double.nan;
                }
                else if ( line == '-Infinity' )
                {
                    return double.negativeInfinity;
                }
                else if ( line == 'Infinity' )
                {
                    return double.infinity;
                }
                else if ( _hexadecimalIntegerExpression.hasMatch( line ) ||
                        _decimalRealExpression.hasMatch( line ) ||
                        _exponentialDecimalRealExpression.hasMatch( line ) )
                {
                    num? number = num.tryParse( line );

                    if ( number != null
                         && !number.isNaN )
                    {
                        return number;
                    }
                }
            }

            context.lineIndex--;

            return parseDefUnquotedString( context, level );
        }
    }
    else
    {
        throwParsingError( 'Invalid DEF value', context, level );
    }
}

// ~~

dynamic parseDefText(
    String text,
    {
        String baseFolderPath = '',
        String filePath = '',
        String Function( String, [String] )? fileReadingFunction = null,
        String stringProcessingQuote = '\'',
        dynamic Function( String, ParsingContext, int )? stringProcessingFunction = processDefQuotedString,
        int levelSpaceCount = 4
    }
    )
{
    var lineArray =
        text
            .replaceAll( '\t', ' ' * levelSpaceCount )
            .replaceAll( '\r', '' )
            .split( '\n' );

    var context =
        ParsingContext(
            baseFolderPath: baseFolderPath,
            filePath: filePath,
            fileReadingFunction: fileReadingFunction,
            stringProcessingQuote: stringProcessingQuote,
            stringProcessingFunction: stringProcessingFunction,
            levelSpaceCount: levelSpaceCount,
            text: text,
            lineArray: lineArray
            );

    return parseDefValue( context, 0 );
}

File: CODE/PYTHON/TEXT/parsing.py
# -- IMPORTS

import re;

from .constant import undefined;
from .map import Map;
from .processing import process_def_quoted_string;

# -- CONSTANTS

decimal_real_expression = re.compile( r"^-?\d+(\.\d+)?$" );
exponential_decimal_real_expression = re.compile( r"^-?\d+(\.\d+)?[eE][-+]?\d+$" );
hexadecimal_integer_expression = re.compile( r"^0x[0-9A-Fa-f]+$" );

# -- FUNCTIONS

def get_key_text(
    key
    ):

    if key is not None and isinstance( key, dict ):

        import json;

        return json.dumps( key );

    else:

        return str( key );

# ~~

def get_token_array(
    text
    ):

    token_array = [];
    character_index = 0;

    while character_index < len( text ):

        if text[ character_index ] == "\\":

            if character_index + 1 < len( text ):

                token_array.append( text[ character_index ] + text[ character_index + 1 ] );
                character_index += 2;
            else:

                token_array.append( text[ character_index ] );
                character_index += 1;

        else:

            post_character_index = character_index;

            while post_character_index < len( text ) and text[ post_character_index ] != "\\":

                post_character_index += 1;

            token_array.append( text[ character_index:post_character_index ] );
            character_index = post_character_index;

    return token_array;

# ~~

def get_unescaped_text(
    token_array
    ):

    unescaped_text = "";

    for token in token_array:

        if len( token ) == 2 and token[ 0 ] == "\\":

            if token[ 1 ] == "n":

                unescaped_text += "\n";

            elif token[ 1 ] == "t":

                unescaped_text += "\t";

            elif token[ 1 ] == "r":

                unescaped_text += "\r";

            elif token[ 1 ] == "b":

                unescaped_text += "\b";

            elif token[ 1 ] == "f":

                unescaped_text += "\f";

            elif token[ 1 ] == "0":

                unescaped_text += "\0";

            else:

                unescaped_text += token[ 1 ];

        else:

            unescaped_text += token;

    return unescaped_text;

# ~~

def get_object_key_text( key ):

    if key is undefined:

        return "undefined";

    if key is None:

        return "null";

    if key is True:

        return "true";

    if key is False:

        return "false";

    if isinstance( key, float ):

        if key != key:

            return "NaN";

        if key == float( "inf" ):

            return "Infinity";

        if key == float( "-inf" ):

            return "-Infinity";

    return str( key );

# ~~

def throw_parsing_error(
    message,
    context,
    level
    ):

    message = (
        message
        + "\nText :\n"
        + context[ "text" ]
        + "\nFile : "
        + context[ "file_path" ]
        + "\nLine "
        + str( context[ "line_index" ] )
        + " @ "
        + str( level )
        );

    if context[ "line_index" ] > 0 and context[ "line_index" ] <= len( context[ "line_array" ] ):

        message += " : " + context[ "line_array" ][ context[ "line_index" ] - 1 ];

    raise Exception( message );

# ~~

def parse_def_line(
    context,
    level
    ):

    line = context[ "line_array" ][ context[ "line_index" ] ];
    trimmed_line = line.lstrip();
    level_space_count = level * context[ "level_space_count" ];
    line_space_count = len( line ) - len( trimmed_line );

    if trimmed_line == "":

        line = "";
        line_space_count = 0;

    else:

        if line_space_count < level_space_count:

            throw_parsing_error( "Invalid DEF line", context, level );

        line = line[ level_space_count: ].rstrip();
        line_space_count -= level_space_count;

    context[ "line_index" ] += 1;

    return { "line": line, "line_space_count": line_space_count };

# ~~

def parse_def_unquoted_string(
    context,
    level
    ):

    string = "";

    while context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        token_array = get_token_array( line );
        last_token = token_array[ len( token_array ) - 1 ] if len( token_array ) > 0 else "";

        if last_token == "\\":

            token_array.pop();

            string += get_unescaped_text( token_array );

        else:

            if line.endswith( "¨" ) and last_token != "\\¨":

                token_array[ len( token_array ) - 1 ] = last_token[ : len( last_token ) - 1 ];

            string += get_unescaped_text( token_array );

            return string;

    throw_parsing_error( "Invalid DEF unquoted string", context, level );

# ~~

def parse_def_quoted_string(
    context,
    level
    ):

    first_line_index = context[ "line_index" ] + 1;
    string = "";
    quote = "";
    escaped_quote = "";

    while context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        if context[ "line_index" ] == first_line_index:

            quote = line[ 0 ];
            line = line[ 1: ];
            escaped_quote = "\\" + quote;

        token_array = get_token_array( line );
        last_token = token_array[ len( token_array ) - 1 ] if len( token_array ) > 0 else "";

        if last_token == "\\":

            token_array.pop();

            string += get_unescaped_text( token_array );

        elif last_token.endswith( quote ) and last_token != escaped_quote:

            token_array[ len( token_array ) - 1 ] = last_token[ : len( last_token ) - 1 ];

            string += get_unescaped_text( token_array );

            if quote == context[ "string_processing_quote" ] and context[ "process_quoted_string_function" ] is not None:

                return context[ "process_quoted_string_function" ]( string, context, level );

            else:

                return string;

        else:

            if last_token.endswith( "¨" ) and last_token != "\\¨":

                token_array[ len( token_array ) - 1 ] = last_token[ : len( last_token ) - 1 ];

            string += get_unescaped_text( token_array ) + "\n";

    throw_parsing_error( "Invalid DEF quoted string", context, level );

# ~~

def parse_def_array(
    context,
    level
    ):

    array = [];

    while context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        if line_space_count == 0 and line == "]":

            return array;

        elif line_space_count == 0 and line == "]{}":

            if len( array ) == 0:

                return [];

            else:

                if isinstance( array[ 0 ], list ):

                    if len( array ) == 1:

                        return [];

                    else:

                        key_array = [];

                        for key in array[ 0 ]:

                            key_array.append( get_key_text( key ) );

                        key_count = len( key_array );
                        object_array = [];
                        object_count = len( array ) - 1;

                        for object_index in range( object_count ):

                            value_array = array[ object_index + 1 ];

                            if isinstance( value_array, list ) and len( value_array ) == key_count:

                                obj = {};

                                for key_index in range( key_count ):

                                    obj[ key_array[ key_index ] ] = value_array[ key_index ];

                                object_array.append( obj );

                            else:

                                throw_parsing_error( "Invalid DEF object array", context, level );

                        return object_array;
                else:

                    throw_parsing_error( "Invalid DEF object array", context, level );

        elif line_space_count == 0 and line == "]()":

            if len( array ) == 0:

                return [];

            else:

                if isinstance( array[ 0 ], list ):

                    if len( array ) == 1:

                        return [];

                    else:

                        key_array = [];

                        for key in array[ 0 ]:

                            key_array.append( get_key_text( key ) );

                        key_count = len( key_array );
                        map_array = [];
                        map_count = len( array ) - 1;

                        for map_index in range( map_count ):

                            value_array = array[ map_index + 1 ];

                            if isinstance( value_array, list ) and len( value_array ) == key_count:

                                map_value = Map();

                                for key_index in range( key_count ):

                                    map_value.set( key_array[ key_index ], value_array[ key_index ] );

                                map_array.append( map_value );

                            else:

                                throw_parsing_error( "Invalid DEF map array", context, level );

                        return map_array;

                else:

                    throw_parsing_error( "Invalid DEF map array", context, level );

        else:

            context[ "line_index" ] -= 1;

            value = parse_def_value( context, level + 1 );
            array.append( value );

    throw_parsing_error( "Invalid DEF array", context, level );

# ~~

def parse_def_object(
    context,
    level
    ):

    obj = {};

    while context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        if line_space_count == 0 and line == "}":

            return obj;

        else:

            context[ "line_index" ] -= 1;

            key = parse_def_value( context, level + 1 );
            value = parse_def_value( context, level + 2 );

            if key is not None and isinstance( key, dict ):

                import json;

                obj[ json.dumps( key ) ] = value;

            else:

                obj[ get_object_key_text( key ) ] = value;

    throw_parsing_error( "Invalid DEF object", context, level );

# ~~

def parse_def_map(
    context,
    level
    ):

    map_value = Map();

    while context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        if line_space_count == 0 and line == ")":

            return map_value;

        else:

            context[ "line_index" ] -= 1;

            key = parse_def_value( context, level + 1 );
            value = parse_def_value( context, level + 2 );

            map_value.set( key, value );

    throw_parsing_error( "Invalid DEF map", context, level );

# ~~

def parse_def_value(
    context,
    level
    ):

    if context[ "line_index" ] < len( context[ "line_array" ] ):

        parsed_line = parse_def_line( context, level );
        line = parsed_line[ "line" ];
        line_space_count = parsed_line[ "line_space_count" ];

        if line == "[":

            return parse_def_array( context, level );

        elif line == "{":

            return parse_def_object( context, level );

        elif line == "(":

            return parse_def_map( context, level );

        elif (
            line.startswith( "'" )
            or line.startswith( '"' )
            or line.startswith( "`" )
            or line.startswith( "´" )
            ):

            context[ "line_index" ] -= 1;

            return parse_def_quoted_string( context, level );

        else:

            token_array = get_token_array( line );

            if len( token_array ) == 1 and token_array[ 0 ] == line.strip():

                if line == "undefined":

                    return undefined;

                elif line == "null":

                    return None;

                elif line == "false":

                    return False;

                elif line == "true":

                    return True;

                elif line == "NaN":

                    return float( "nan" );

                elif line == "-Infinity":

                    return float( "-inf" );

                elif line == "Infinity":

                    return float( "inf" );

                elif hexadecimal_integer_expression.match( line ):

                    try:

                        return int( line, 0 );

                    except ValueError:

                        pass;

                elif decimal_real_expression.match( line ) or exponential_decimal_real_expression.match( line ):

                    try:

                        number = float( line );

                        if number == int( number ):

                            return int( number );
                        else:

                            return number;

                    except ValueError:

                        pass;

            context[ "line_index" ] -= 1;

            return parse_def_unquoted_string( context, level );

    else:

        throw_parsing_error( "Invalid DEF value", context, level );

# ~~

def parse_def_text(
    text,
    options=None
    ):

    if options is None:

        options = {};

    base_folder_path = options.get( "base_folder_path", "" );
    file_path = options.get( "file_path", "" );
    read_text_file_function = options.get( "read_text_file_function", None );
    string_processing_quote = options.get( "string_processing_quote", "'" );
    process_quoted_string_function = options.get( "process_quoted_string_function", process_def_quoted_string );
    level_space_count = options.get( "level_space_count", 4 );

    line_array = (
        text.rstrip()
        .replace( "\t", " " * level_space_count )
        .replace( "\r", "" )
        .split( "\n" )
        );

    context = {
        "base_folder_path": base_folder_path,
        "file_path": file_path,
        "read_text_file_function": read_text_file_function,
        "string_processing_quote": string_processing_quote,
        "process_quoted_string_function": process_quoted_string_function,
        "level_space_count": level_space_count,
        "text": text,
        "line_array": line_array,
        "line_index": 0
        };

    return parse_def_value( context, 0 );
